index.js

```diff
@@ -125,6 +145,172 @@
     return result;
   };
 
+  var sigGroup = function(offset, separator, glue) {
+    return {
+      prefix: '',
+      suffix: '',
+      glue: glue,
+      separator: separator,
+      start: offset,
+      end: offset,
+      items: []
+    };
+  };
+  
+  var sigItem = function(type, offset, text, glue) {
+    return {
+      type: type,
+      text: text,
+      glue: glue,
+      start: offset,
+      end: offset + text.length,
+      children: [],
+      typeVars: []
+    };
+  };
+  
+  var sigTypeVar = function(text, offset) {
+    return {
+      text: text,
+      offset: offset
+    };
+  };
+
+  var sigName = function(name, offset) {
+    var group = sigGroup(offset, ' :: ', '');
+    group.end = name.length + group.separator.length;
+    group.items.push(sigItem('__NAME__', offset, name, ''));
+    return group;
+  };
+  
+  var sigConstraints = function(constraints, offset) {
+    var group = sigGroup(offset, ' => ', ', ');
+    return group;
+  };
+  
+  var sigArguments = function(expTypes, offset) {
+    var group = sigGroup(offset, ' -> ', ' -> ');
+    if (isEmpty(keys(expTypes))) {
+      return group;
+    }
+    else {
+      var idx, item;
+      for (idx = 0; idx < expTypes.length; ++idx) {
+        if (idx > 0) {
+          group.end += group.glue.length;
+        }
+        item = sigArgument(expTypes[idx], group.end, false);
+        group.items.push(item);
+        group.end = item.end;
+      }
+      return group;
+    }
+  };
+  
+  var isContainerType = function(t) {
+    return t.type == 'UNARY' || t.type == 'BINARY';
+  }
+  
+  var sigArgument = function(t, offset, isChild) {
+    var item = sigItem(t.type, offset,
+                       t.name ? stripNamespace(t.name) : t.toString(), ' ');
+    if (t.type == 'VARIABLE') {
+      item.typeVars[t.name] = sigTypeVar(t.name, offset);
+    } else if (isContainerType(t)) {
+      if (isChild) {
+        item.text = '(' + item.text;
+        item.end += 1;
+      }
+      var childTypes = [t.$1];
+      if (t.type == 'BINARY') {
+        childTypes.push(t.$2);
+      }
+      var idx, child;
+      for (idx = 0; idx < childTypes.length; idx += 1) {
+        item.end += item.glue.length;
+        child = sigArgument(childTypes[idx], item.end, true);
+        if (isContainerType(child)) {
+          item.children.push(child);
+        } else {
+          item.typeVars = merge(item.typeVars)(child.typeVars);
+        }
+        item.text += item.glue + child.text;
+        item.end = offset + item.text.length;
+      }
+      if (isChild) {
+        item.text += ')';
+        item.end += 1;
+      }
+    }
+    return item;
+  };
```

Comment:
Recursively creating the argument [items](https://github.com/plaid/sanctuary-def/pull/23/files#diff-168726dbe96b3ce427e7fedce31bb0bcR160). I was planning to change this to make two passes, one to create the text and another to compute the offsets.. thoughts? We hardly need to worry about perf here at all since the sig structure is only built once and only when there is an error to be thrown. We could even make multiple passes just to create the text, ie don't work about wrapping children in parentheses as we go and do it in a 2nd pass.